File: rust/src/issue_listing.rs
//! Issue listing utilities.

use std::io::ErrorKind;
use std::path::Path;

use crate::config_loader::load_project_configuration;
use crate::daemon_client::{is_daemon_enabled, request_index_list};
use crate::error::KanbusError;
use crate::file_io::{
    canonicalize_path, discover_kanbus_projects, discover_project_directories,
    find_project_local_directory, get_configuration_path, load_project_directory,
    resolve_labeled_projects,
};
use crate::models::IssueData;
use crate::queries::{filter_issues, search_issues, sort_issues};
use std::collections::HashSet;

/// List issues for the project.
///
/// # Arguments
/// * `root` - Repository root path.
///
/// # Errors
/// Returns `KanbusError` when listing fails.
#[allow(clippy::too_many_arguments)]
pub fn list_issues(
    root: &Path,
    status: Option<&str>,
    issue_type: Option<&str>,
    assignee: Option<&str>,
    label: Option<&str>,
    sort: Option<&str>,
    search: Option<&str>,
    project_filter: &[String],
    include_local: bool,
    local_only: bool,
) -> Result<Vec<IssueData>, KanbusError> {
    if local_only && !include_local {
        return Err(KanbusError::IssueOperation(
            "local-only conflicts with no-local".to_string(),
        ));
    }
    if !project_filter.is_empty() {
        return list_with_project_filter(
            root,
            project_filter,
            status,
            issue_type,
            assignee,
            label,
            sort,
            search,
            include_local,
            local_only,
        );
    }
    let mut projects = Vec::new();
    discover_project_directories(root, &mut projects)?;
    let mut dotfile_projects = discover_kanbus_projects(root)?;
    projects.append(&mut dotfile_projects);
    let mut normalized = Vec::new();
    for path in projects {
        match canonicalize_path(&path) {
            Ok(canonical) => normalized.push(canonical),
            Err(_) => normalized.push(path),
        }
    }
    normalized.sort();
    normalized.dedup();
    if let Ok(config_path) = get_configuration_path(root) {
        if let Ok(configuration) = load_project_configuration(&config_path) {
            let base = config_path.parent().unwrap_or_else(|| Path::new(""));
            normalized.retain(|project_path| {
                !crate::file_io::is_path_ignored(project_path, base, &configuration.ignore_paths)
            });
        }
    }
    let mut permission_error = None;
    normalized.retain(|path| {
        let issues_dir = path.join("issues");
        match std::fs::metadata(&issues_dir) {
            Ok(metadata) => metadata.is_dir(),
            Err(error) => {
                if error.kind() == ErrorKind::PermissionDenied {
                    permission_error = Some(error);
                }
                false
            }
        }
    });
    if let Some(error) = permission_error {
        return Err(KanbusError::Io(error.to_string()));
    }
    let projects = normalized;
    if projects.is_empty() {
        return Err(KanbusError::IssueOperation(
            "project not initialized".to_string(),
        ));
    }
    if projects.len() > 1 {
        let issues = list_issues_across_projects(root, &projects, include_local, local_only)?;
        return apply_query(issues, status, issue_type, assignee, label, sort, search);
    }

    if include_local || local_only {
        let project_dir = load_project_directory(root)?;
        let local_dir = find_project_local_directory(&project_dir);
        if !local_only && is_daemon_enabled() {
            let payloads = request_index_list(root)?;
            let mut issues: Vec<IssueData> = payloads
                .into_iter()
                .map(serde_json::from_value::<IssueData>)
                .map(|result| result.map_err(|error| KanbusError::Io(error.to_string())))
                .collect::<Result<Vec<IssueData>, KanbusError>>()?;
            if let Some(local_dir) = local_dir {
                let local_issues_dir = local_dir.join("issues");
                if local_issues_dir.exists() {
                    issues.extend(load_issues_from_directory(&local_issues_dir)?);
                }
            }
            return apply_query(issues, status, issue_type, assignee, label, sort, search);
        }
        let issues = list_issues_with_local(&project_dir, local_dir.as_deref(), local_only)?;
        return apply_query(issues, status, issue_type, assignee, label, sort, search);
    }
    if is_daemon_enabled() {
        let payloads = request_index_list(root)?;
        let issues: Vec<IssueData> = payloads
            .into_iter()
            .map(serde_json::from_value::<IssueData>)
            .map(|result| result.map_err(|error| KanbusError::Io(error.to_string())))
            .collect::<Result<Vec<IssueData>, KanbusError>>()?;
        return apply_query(issues, status, issue_type, assignee, label, sort, search);
    }
    let issues = list_issues_local(root)?;
    apply_query(issues, status, issue_type, assignee, label, sort, search)
}

#[allow(clippy::too_many_arguments)]
fn list_with_project_filter(
    root: &Path,
    project_filter: &[String],
    status: Option<&str>,
    issue_type: Option<&str>,
    assignee: Option<&str>,
    label: Option<&str>,
    sort: Option<&str>,
    search: Option<&str>,
    include_local: bool,
    local_only: bool,
) -> Result<Vec<IssueData>, KanbusError> {
    let labeled = resolve_labeled_projects(root)?;
    if labeled.is_empty() {
        return Err(KanbusError::IssueOperation(
            "project not initialized".to_string(),
        ));
    }
    let known: HashSet<&str> = labeled.iter().map(|p| p.label.as_str()).collect();
    for name in project_filter {
        if !known.contains(name.as_str()) {
            return Err(KanbusError::IssueOperation(format!(
                "unknown project: {name}"
            )));
        }
    }
    let allowed: HashSet<&str> = project_filter.iter().map(|s| s.as_str()).collect();
    let project_dirs: Vec<std::path::PathBuf> = labeled
        .into_iter()
        .filter(|p| allowed.contains(p.label.as_str()))
        .map(|p| p.project_dir)
        .collect();
    let issues = list_issues_across_projects(root, &project_dirs, include_local, local_only)?;
    apply_query(issues, status, issue_type, assignee, label, sort, search)
}

fn list_issues_local(root: &Path) -> Result<Vec<IssueData>, KanbusError> {
    let project_dir = load_project_directory(root)?;
    list_issues_for_project(&project_dir)
}

fn list_issues_for_project(project_dir: &Path) -> Result<Vec<IssueData>, KanbusError> {
    let issues_dir = project_dir.join("issues");
    if !issues_dir.is_dir() {
        return Err(KanbusError::IssueOperation(format!(
            "issues directory not found: {}. Run 'kanbus migrate' if you need to migrate from an older format.",
            issues_dir.display()
        )));
    }
    load_issues_from_directory(&issues_dir)
}

fn list_issues_with_local(
    project_dir: &Path,
    local_dir: Option<&Path>,
    local_only: bool,
) -> Result<Vec<IssueData>, KanbusError> {
    if std::env::var("KANBUS_TEST_LOCAL_LISTING_ERROR").is_ok() {
        return Err(KanbusError::IssueOperation(
            "local listing failed".to_string(),
        ));
    }
    let shared_issues = list_issues_for_project(project_dir)?;
    let mut local_issues = Vec::new();
    if let Some(local_dir) = local_dir {
        let issues_dir = local_dir.join("issues");
        if issues_dir.exists() {
            local_issues = load_issues_from_directory(&issues_dir)?;
        }
    }
    if local_only {
        return Ok(local_issues);
    }
    Ok([shared_issues, local_issues].concat())
}

fn list_issues_across_projects(
    root: &Path,
    projects: &[std::path::PathBuf],
    include_local: bool,
    local_only: bool,
) -> Result<Vec<IssueData>, KanbusError> {
    let mut issues = Vec::new();
    for project_dir in projects {
        let issues_dir = project_dir.join("issues");
        if !issues_dir.is_dir() {
            continue;
        }
        let local_dir = if include_local || local_only {
            find_project_local_directory(project_dir)
        } else {
            None
        };
        if local_only && local_dir.is_none() {
            continue;
        }
        let mut project_issues =
            list_issues_with_local(project_dir, local_dir.as_deref(), local_only)?;
        for issue in &mut project_issues {
            tag_issue_project(issue, root, project_dir);
        }
        issues.extend(project_issues);
    }
    Ok(issues)
}

fn tag_issue_project(issue: &mut IssueData, root: &Path, project_dir: &Path) {
    let project_path = project_dir
        .strip_prefix(root)
        .map(|path| path.to_path_buf())
        .unwrap_or_else(|_| project_dir.to_path_buf());
    issue.custom.insert(
        "project_path".to_string(),
        serde_json::Value::String(project_path.to_string_lossy().to_string()),
    );
}

fn load_issues_from_directory(issues_dir: &Path) -> Result<Vec<IssueData>, KanbusError> {
    let mut issues = Vec::new();
    for entry in
        std::fs::read_dir(issues_dir).map_err(|error| KanbusError::Io(error.to_string()))?
    {
        let entry = entry.map_err(|error| KanbusError::Io(error.to_string()))?;
        let path = entry.path();
        if path.extension().and_then(|ext| ext.to_str()) != Some("json") {
            continue;
        }
        issues.push(crate::issue_files::read_issue_from_file(&path)?);
    }
    issues.sort_by(|left, right| left.identifier.cmp(&right.identifier));
    Ok(issues)
}

fn apply_query(
    issues: Vec<IssueData>,
    status: Option<&str>,
    issue_type: Option<&str>,
    assignee: Option<&str>,
    label: Option<&str>,
    sort: Option<&str>,
    search: Option<&str>,
) -> Result<Vec<IssueData>, KanbusError> {
    let filtered = filter_issues(issues, status, issue_type, assignee, label);
    let searched = search_issues(filtered, search);
    sort_issues(searched, sort)
}

File: python/src/kanbus/issue_listing.py
"""Issue listing utilities."""

from __future__ import annotations

from pathlib import Path
from typing import List

from kanbus.cache import collect_issue_file_mtimes, load_cache_if_valid, write_cache
from kanbus.daemon_client import is_daemon_enabled, request_index_list
from kanbus.index import build_index_from_directory
from kanbus.issue_files import read_issue_from_file
from kanbus.models import IssueData
from kanbus.project import (
    ProjectMarkerError,
    ResolvedProject,
    discover_project_directories,
    find_project_local_directory,
    load_project_directory,
    resolve_labeled_projects,
    resolve_project_path,
)
from kanbus.migration import MigrationError, load_beads_issues
from kanbus.queries import filter_issues, search_issues, sort_issues


class IssueListingError(RuntimeError):
    """Raised when listing issues fails."""


def list_issues(
    root: Path,
    status: str | None = None,
    issue_type: str | None = None,
    assignee: str | None = None,
    label: str | None = None,
    sort: str | None = None,
    search: str | None = None,
    project_filter: List[str] | None = None,
    include_local: bool = True,
    local_only: bool = False,
    beads_mode: bool = False,
) -> List[IssueData]:
    """List issues in the project.

    :param root: Repository root path.
    :type root: Path
    :return: List of issues.
    :rtype: List[IssueData]
    :param project_filter: Optional list of project labels to include.
    :type project_filter: List[str] | None
    :param beads_mode: Whether to read from Beads JSONL instead of project files.
    :type beads_mode: bool
    :raises IssueListingError: If listing fails.
    """
    if local_only and not include_local:
        raise IssueListingError("local-only conflicts with no-local")
    if beads_mode:
        if local_only or not include_local:
            raise IssueListingError("beads mode does not support local filtering")
        try:
            issues = load_beads_issues(root)
        except MigrationError as error:
            raise IssueListingError(str(error)) from error
        # Default: exclude closed issues unless status is explicitly specified
        if status is None:
            issues = [issue for issue in issues if issue.status != "closed"]
        return _apply_query(issues, status, issue_type, assignee, label, sort, search)

    if project_filter:
        return _list_with_project_filter(
            root, project_filter, status, issue_type, assignee, label,
            sort, search, include_local, local_only,
        )

    try:
        project_dirs = discover_project_directories(root)
    except ProjectMarkerError as error:
        raise IssueListingError(str(error)) from error
    if not project_dirs:
        raise IssueListingError("project not initialized")

    if len(project_dirs) > 1:
        issues = _list_issues_across_projects(
            root, project_dirs, include_local, local_only
        )
        return _apply_query(issues, status, issue_type, assignee, label, sort, search)

    project_dir = project_dirs[0]
    local_dir = None
    if include_local or local_only:
        local_dir = find_project_local_directory(project_dir)

    if local_only:
        try:
            issues = _list_issues_with_local(
                project_dir,
                local_dir,
                include_local,
                local_only,
            )
            return _apply_query(
                issues, status, issue_type, assignee, label, sort, search
            )
        except Exception as error:
            raise IssueListingError(str(error)) from error

    shared_issues: List[IssueData]
    if is_daemon_enabled():
        try:
            payloads = request_index_list(root)
            shared_issues = [IssueData.model_validate(payload) for payload in payloads]
            shared_issues = [
                _tag_issue_source(issue, "shared") for issue in shared_issues
            ]
        except Exception as error:
            raise IssueListingError(str(error)) from error
    else:
        try:
            shared_issues = _list_issues_locally(root)
            shared_issues = [
                _tag_issue_source(issue, "shared") for issue in shared_issues
            ]
        except Exception as error:
            raise IssueListingError(str(error)) from error

    if include_local and local_dir is not None:
        try:
            issues_dir = local_dir / "issues"
            local_issues: List[IssueData] = []
            if issues_dir.exists():
                local_issues = [
                    _tag_issue_source(issue, "local")
                    for issue in _load_issues_from_directory(issues_dir)
                ]
            shared_issues = [*shared_issues, *local_issues]
        except Exception as error:
            raise IssueListingError(str(error)) from error

    return _apply_query(
        shared_issues, status, issue_type, assignee, label, sort, search
    )


def _list_with_project_filter(
    root: Path,
    project_filter: List[str],
    status: str | None,
    issue_type: str | None,
    assignee: str | None,
    label: str | None,
    sort: str | None,
    search: str | None,
    include_local: bool,
    local_only: bool,
) -> List[IssueData]:
    try:
        labeled = resolve_labeled_projects(root)
    except ProjectMarkerError as error:
        raise IssueListingError(str(error)) from error
    if not labeled:
        raise IssueListingError("project not initialized")
    known_labels = {project.label for project in labeled}
    for name in project_filter:
        if name not in known_labels:
            raise IssueListingError(f"unknown project: {name}")
    allowed = set(project_filter)
    filtered_projects = [project for project in labeled if project.label in allowed]
    project_dirs = [project.project_dir for project in filtered_projects]
    issues = _list_issues_across_projects(root, project_dirs, include_local, local_only)
    return _apply_query(issues, status, issue_type, assignee, label, sort, search)


def _list_issues_locally(root: Path) -> List[IssueData]:
    project_dir = load_project_directory(root)
    return _list_issues_for_project(project_dir)


def _list_issues_for_project(project_dir: Path) -> List[IssueData]:
    issues_dir = project_dir / "issues"
    if not issues_dir.is_dir():
        # Beads fallback: project_dir is typically <repo>/project, so
        # the repo root is one level up.
        repo_root = project_dir.parent
        beads_path = repo_root / ".beads" / "issues.jsonl"
        if beads_path.exists():
            try:
                return load_beads_issues(repo_root)
            except MigrationError:
                return []
        return []
    cache_path = project_dir / ".cache" / "index.json"
    cached = load_cache_if_valid(cache_path, issues_dir)
    if cached is not None:
        return list(cached.by_id.values())
    index = build_index_from_directory(issues_dir)
    mtimes = collect_issue_file_mtimes(issues_dir)
    write_cache(index, cache_path, mtimes)
    return list(index.by_id.values())


def _list_issues_with_local(
    project_dir: Path,
    local_dir: Path | None,
    include_local: bool,
    local_only: bool,
) -> List[IssueData]:
    shared_issues = _list_issues_for_project(project_dir)
    shared_tagged = [_tag_issue_source(issue, "shared") for issue in shared_issues]

    local_tagged: List[IssueData] = []
    if local_dir is not None:
        issues_dir = local_dir / "issues"
        if issues_dir.exists():
            local_tagged = [
                _tag_issue_source(issue, "local")
                for issue in _load_issues_from_directory(issues_dir)
            ]

    if local_only:
        return local_tagged
    if include_local:
        return [*shared_tagged, *local_tagged]
    return shared_tagged


def _list_issues_across_projects(
    root: Path,
    project_dirs: List[Path],
    include_local: bool,
    local_only: bool,
) -> List[IssueData]:
    issues: List[IssueData] = []
    for project_dir in sorted(project_dirs):
        local_dir = None
        if include_local or local_only:
            local_dir = find_project_local_directory(project_dir)
        if local_only and local_dir is None:
            continue
        project_issues = _list_issues_with_local(
            project_dir,
            local_dir,
            include_local,
            local_only,
        )
        project_issues = [
            _tag_issue_project(issue, root, project_dir) for issue in project_issues
        ]
        issues.extend(project_issues)
    return issues


def _load_issues_from_directory(issues_dir: Path) -> List[IssueData]:
    issues = [
        read_issue_from_file(path)
        for path in sorted(issues_dir.glob("*.json"), key=lambda item: item.name)
    ]
    return issues


def _tag_issue_source(issue: IssueData, source: str) -> IssueData:
    custom = {**issue.custom, "source": source}
    return issue.model_copy(update={"custom": custom})


def _tag_issue_project(issue: IssueData, root: Path, project_dir: Path) -> IssueData:
    project_path = _render_project_path(root, project_dir)
    custom = {**issue.custom, "project_path": project_path}
    return issue.model_copy(update={"custom": custom})


def _render_project_path(root: Path, project_dir: Path) -> str:
    root_resolved = resolve_project_path(root)
    project_resolved = resolve_project_path(project_dir)
    try:
        project_path = project_resolved.relative_to(root_resolved)
    except ValueError:
        project_path = project_resolved
    return str(project_path)


def _apply_query(
    issues: List[IssueData],
    status: str | None,
    issue_type: str | None,
    assignee: str | None,
    label: str | None,
    sort: str | None,
    search: str | None,
) -> List[IssueData]:
    filtered = filter_issues(issues, status, issue_type, assignee, label)
    searched = search_issues(filtered, search)
    return sort_issues(searched, sort)
